src/rust.rs

```diff
@@ -657,40 +668,90 @@ macro_rules! auto_root {
     }
 }
 
-impl<T> Handle<T> {
+pub struct Handle<'a, T: 'a> {
+    ptr: &'a T,
+}
+
+pub struct MutableHandle<'a, T: 'a> {
+    ptr: &'a mut T
+}
+
+pub type MutableHandleValue<'a> = MutableHandle<'a, Value>;
+
+pub type HandleValue<'a> = Handle<'a, Value>;
+pub type HandleObject<'a> = Handle<'a, *mut JSObject>;
+pub type HandleId<'a> = Handle<'a, jsid>;
+
+impl<T> RawHandle<T> {
     pub fn get(&self) -> T
         where T: Copy
     {
         unsafe { *self.ptr }
     }
 
-    pub unsafe fn from_marked_location(ptr: *const T) -> Handle<T> {
-        Handle {
+    pub unsafe fn from_marked_location(ptr: *const T) -> Self {
+        RawHandle {
             _base: HandleBase { _phantom0: PhantomData },
             ptr: ptr,
         }
     }
 }
 
-impl<T> Deref for Handle<T> {
+impl<'a, T> Handle<'a, T> {
+    pub fn get(&self) -> T
+        where T: Copy
+    {
+        *self.ptr
+    }
+
+    pub fn new(ptr: &'a T) -> Self {
+        Handle { ptr: ptr }
+    }
+
+    pub unsafe fn from_marked_location(ptr: *const T) -> Self {
```

Comment:
Is this used outside of this crate? If we could avoid making it public, that would be useful.